api/models/tables.py

```diff
@@ -139,6 +145,92 @@ def raw_data_for_geos(self, geos):
+        :param bool percent: should we calculate percentages, or just include raw values?
+        :param int total: the total value to use for percentages, name of a
+                          field, or None to use the sum of all retrieved fields (default)
+        :param dict recode: map from field names to strings to recode column names. Many fields
+                            can be recoded to the same thing, their values will be summed.
+
+        :return: (data-dictionary, total)
+        """
+
+        session = get_session()
+        try:
+            if fields is not None and not isinstance(fields, list):
+                fields = [fields]
+            if fields:
+                for f in fields:
+                    if f not in self.columns:
+                        raise ValueError('Invalid field/column for %s: %s' % (self.id, f))
+            else:
+                fields = self.columns.keys()
+
+            recode = recode or {}
+            if recode:
+                # change lambda to dicts
+                if not isinstance(recode, dict):
+                    recode = {f: recode(f) for f in fields}
+
+            # table columns to fetch
+            cols = [self.model.columns[c] for c in fields]
+            if total is not None and isinstance(total, basestring) and total not in cols:
+                cols.append(total)
+
+            # do the query. If this returns no data, row is None
+            row = session\
+                .query(*cols)\
+                .filter(self.model.c.geo_level == geo_level,
+                        self.model.c.geo_code == geo_code)\
+                .first()
+
+            if row is None:
+                row = ZeroRow()
+
+            # what's our denominator?
+            if total is None:
+                # sum of all columns
+                total = sum(getattr(row, f) or 0 for f in fields)
+            elif isinstance(total, basestring):
+                total = getattr(row, total)
+
+            results = OrderedDict()
+            key_order = key_order or fields
+            for field in key_order:
+                val = getattr(row, field) or 0
+
+                # recode the key, default is to keep it the same
+                key = recode.get(field, field)
+
+                # key may already exist if another column recoded to it
+                results.setdefault(key, {'name': recode.get(field, self.columns[field]['name'])})
+
+                if percent:
+                    # sum up existing values, if any
+                    val = val + results.get('numerators', {}).get('this', 0)
+                    results[key]['values'] = {'this': p(val, total)}
+                    results[key]['numerators'] = {'this': val}
+                else:
+                    # sum up existing values, if any
+                    val = val + results.get('values', {}).get('this', 0)
+                    results[key]['values'] = {'this': val}
```

Comment:
I'm not sure I understand what is happening here. Could we maybe have a quick chat about this a bit later?